### ISSUE TITLE
cannot build working DLL for Unity Windows standalone

### ISSUE THREAD

**Issue by turowskipaul** [libpd/libpd#185]
I cannot seem to build a working DLL for a Windows standalone in Unity at all.  (Building a bundle for Mac works fine)

The prebuilt DLLs on the following pages work fine:
https://github.com/patricksebastien/libpd4unity
https://github.com/Magicolo/uPD
(both use pthreadGC2.dll instead of libwinpthread-1.dll)
 
The problem with these is that they do not include the extra externals and we are using sigmund~ for pitch tracking.
 
I can build my own library without any problems/errors (see below for more details), but when I run the game it crashes immediately and the error report ends on the loading of libpdcsharp.dll without any other helpful information (as far as I can tell).
 
I have carefully reviewed the instructions provided on this page and a number of others, including:
http://bp.io/post/2017
https://github.com/patricksebastien/libpd4unity
http://www.thefuntastic.com/2014/04/the-misadventures-of-unity-and-puredata-libpd/

I've tried:
both versions of msys2 (i686 and x86_64)
both 32- and 64-bit MinGW
both sources for mingw-w64-cross-winpthreads-git (clang and crt).
using legacy versions (2008, 2010) of MinGW (as thefuntastic.com indicates)
rolling back to earlier releases of libpd (0.10.0 and 0.8.3)
both .NET 2.0 and .NET 2.0 subset setting in Unity’s player settings
 
I’ve searched on a number of forums but have not been able to find any solutions that work.

Any suggestions would be much appreciated.

**Comment by residuum** [comment]
Am I reading that correctly, that you can build and use libpd for C# correctly, but you have problems compiling externals for Windows?

**Comment by turowskipaul** [comment]
Thanks for the reply. I'm building with `make UTIL=true EXTRA=true csharplib` from the mingw64 console.  (I've also used the .bat file and had the same results.)  After I compile, I see libpdcsharp.dll in libs (and libwinpthread-1.dll is already in the mingw64 directory) but I don't see sigmund~.dll anywhere, though.  Is that an indication that the EXTRA externals aren't compiling properly?

**Comment by danomatika** [comment]
>I don't see sigmund~.dll anywhere

You won't. The extras are compiled into the lib statically.

**Comment by turowskipaul** [comment]
That was my understanding too.  Also, when I've previously compiled libpd as a bundle and the lib was missing externals used in the Pd patch, the game would still start but I would just get an error that that external was missing.  In the current issue, the game seems to crash immediately after libpdcsharp.dll is loaded (as far as I can tell).

**Comment by danomatika** [comment]
That sounds like an issue with either the wrapper itself or how how the lib is compiled and less likely an issue with compiling the extras or not. Perhaps doing a debug build would at least give meaningful error messages on load? Try setting the `DEBUG=true` makefile option.

**Comment by residuum** [comment]
My bad about the statically compiled extras.

You could open the dll with a text editor and search for the name of the constructor for sigmund~, `sigmund_new`. If it is included, then the symbol should be available.

**Comment by turowskipaul** [comment]
> That sounds like an issue with either the wrapper itself or how how the lib is compiled and less likely an issue with compiling the extras or not.

Yeah it doesn't seem to be related to the extras; compiling libpd without the EXTRA flag also produces a DLL that crashes Unity.

I can build successfully with 'make UTIL=true EXTRA=true DEBUG=true csharplib' using mingw64.  There are a lot of warnings ("cast to pointer from integer of different size...") but nothing that looks problematic as far as I can tell.  I can provide the entire build output if it would help.

**Comment by danomatika** [comment]
>I can provide the entire build output if it would help

I mean more the output when laooding the csharplib, although I suppose Unity would need to be a debug build as well. I am unfamiliar with CSharp land and am extrapolating from C/C++.

**Comment by turowskipaul** [comment]
I am creating a development build in Unity, but it crashes before I can see any console output; immediately after I hit "Play!" I get the splash screen and then a dialog box saying "Oops! The game crashed..."  The crash report generated doesn't seem to offer any helpful info either.  The last lines of error.txt are:
`Module 1
X:\PS_Windows\PluralitySpringv1.1_x86_64_Data\Plugins\libwinpthread-1.dll
Image Base: 0x64940000  Image Size: 0x00086000
File Size:  548081      File Time:  2017-09-01_154358
Version:
   Company:    MingW-W64 Project. All rights reserved.
   Product:    
   FileDesc:   POSIX WinThreads for Windows
   FileVer:    1.0.0.0
   ProdVer:    1.0.0.0

Module 2
X:\PS_Windows\PluralitySpringv1.1_x86_64_Data\Plugins\libpdcsharp.dll
Image Base: 0x66200000  Image Size: 0x0018e000
File Size:  1752088     File Time:  2017-09-01_154358
Version:
   Company:    
   Product:    
   FileDesc:   
   FileVer:    0.0.0.0
   ProdVer:    0.0.0.0


== [end of error.log] ==`

Here's the end of output_log:
`========== OUTPUTING STACK TRACE ==================

RtlLookupFunctionEntry returned NULL function. Aborting stack walk.
0x0000000066282FED (libpdcsharp) gensym
0x00000000662D02A8 (libpdcsharp) libpd_finish_message
0x00000000063413D0 (Mono JIT Code) (wrapper managed-to-native) LibPDBinding.LibPD:finish_message (string,string)
0x0000000006340976 (Mono JIT Code) [/Users/paulturowski/Documents/UnityProjects/PluralitySpring/Assets/LibPD/LibPDNativeMessaging.cs:375] LibPDBinding.LibPD:SendMessage (string,string,object[]) 
0x00000000063408A7 (Mono JIT Code) (wrapper synchronized) LibPDBinding.LibPD:SendMessage (string,string,object[])
0x000000000634081A (Mono JIT Code) [/Users/paulturowski/Documents/UnityProjects/PluralitySpring/Assets/LibPD/LibPDNativeMethods.cs:239] LibPDBinding.LibPD:ComputeAudio (bool) 
0x0000000006340717 (Mono JIT Code) (wrapper synchronized) LibPDBinding.LibPD:ComputeAudio (bool)
0x00000000063403FA (Mono JIT Code) [/Users/paulturowski/Documents/UnityProjects/PluralitySpring/Assets/LibPD/LibPDNativeMethods.cs:208] LibPDBinding.LibPD:Release () 
0x000000000634033F (Mono JIT Code) (wrapper synchronized) LibPDBinding.LibPD:Release ()
0x0000000006340285 (Mono JIT Code) [/Users/paulturowski/Documents/UnityProjects/PluralitySpring/Assets/LibPD/LibPDNativeMethods.cs:80] LibPDBinding.LibPD:ReInit () 
0x000000000634020F (Mono JIT Code) (wrapper synchronized) LibPDBinding.LibPD:ReInit ()
0x000000000633FA0B (Mono JIT Code) [/Users/paulturowski/Documents/UnityProjects/PluralitySpring/Assets/LibPD/LibPDNativeMethods.cs:66] LibPDBinding.LibPD:.cctor () 
0x0000000006320483 (Mono JIT Code) (wrapper runtime-invoke) object:runtime_invoke_void (object,intptr,intptr,intptr)
0x00007FFFE3075B63 (mono) mono_set_defaults
0x00007FFFE2FC872D (mono) mono_runtime_invoke
0x00007FFFE2FCC082 (mono) mono_array_new
0x00007FFFE307529D (mono) mono_set_defaults
0x00007FFFE30754D9 (mono) mono_set_defaults
0x00007FFFE307556C (mono) mono_set_defaults
0x00007FFFE306D643 (mono) mono_debugger_run_finally
<Missing stacktrace information>

========== END OF STACKTRACE ===========

**** Crash! ****`

**Comment by danomatika** [comment]
That's a lot more helpful as it looks like it's a crash inside libpd_finish_message. Is this a build from the current master or from the latest release tag?

**Comment by turowskipaul** [comment]
This is from the current master.  I can revert to the last stable and try again if it would help.

**Comment by danomatika** [comment]
It would help as the current master is in flux based new work from the latest Pd release.

**Comment by turowskipaul** [comment]
When I revert back to release 0.10.0, I'm not even able to compile (using either mingw32 or mingw64. I get errors related to libpdwrapper/z_libpd.c:
`gcc -DPD -DHAVE_UNISTD_H -DUSEAPI_DUMMY -I./pure-data/src -I./libpd_wrapper -I./libpd_wrapper/util -DWINVER=0x502 -DWIN32 -D_WIN32 -DPD_INTERNAL -I"/include" -I"/include/win32" -Wall -DLIBPD_EXTRA    -c -o libpd_wrapper/z_libpd.o libpd_wrapper/z_libpd.c
libpd_wrapper/z_libpd.c: In function 'libpd_init':
libpd_wrapper/z_libpd.c:68:3: error: 'sys_soundin' undeclared (first use in this function)
   sys_soundin = NULL;
   ^~~~~~~~~~~
libpd_wrapper/z_libpd.c:68:3: note: each undeclared identifier is reported only once for each function it appears in
libpd_wrapper/z_libpd.c:69:3: error: 'sys_soundout' undeclared (first use in this function)
   sys_soundout = NULL;
   ^~~~~~~~~~~~
libpd_wrapper/z_libpd.c:71:3: error: 'sys_schedblocksize' undeclared (first use in this function)
   sys_schedblocksize = DEFDACBLKSIZE;
   ^~~~~~~~~~~~~~~~~~
libpd_wrapper/z_libpd.c:78:3: error: 'sys_nogui' undeclared (first use in this function)
   sys_nogui = 1;
   ^~~~~~~~~
libpd_wrapper/z_libpd.c:89:3: error: 'sys_searchpath' undeclared (first use in this function)
   sys_searchpath = NULL;
   ^~~~~~~~~~~~~~
libpd_wrapper/z_libpd.c: In function 'libpd_clear_search_path':
libpd_wrapper/z_libpd.c:109:17: error: 'sys_searchpath' undeclared (first use in this function)
   namelist_free(sys_searchpath);
                 ^~~~~~~~~~~~~~
libpd_wrapper/z_libpd.c: In function 'libpd_add_to_search_path':
libpd_wrapper/z_libpd.c:114:3: error: 'sys_searchpath' undeclared (first use in this function)
   sys_searchpath = namelist_append(sys_searchpath, s, 0);
   ^~~~~~~~~~~~~~
libpd_wrapper/z_libpd.c: In function 'libpd_process_raw':
libpd_wrapper/z_libpd.c:146:17: error: 'sys_inchannels' undeclared (first use in this function)
   size_t n_in = sys_inchannels * DEFDACBLKSIZE;
                 ^~~~~~~~~~~~~~
libpd_wrapper/z_libpd.c:147:18: error: 'sys_outchannels' undeclared (first use in this function)
   size_t n_out = sys_outchannels * DEFDACBLKSIZE;
                  ^~~~~~~~~~~~~~~
libpd_wrapper/z_libpd.c:151:12: error: 'sys_soundin' undeclared (first use in this function)
   for (p = sys_soundin, i = 0; i < n_in; i++) {
            ^~~~~~~~~~~
libpd_wrapper/z_libpd.c:151:23: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   for (p = sys_soundin, i = 0; i < n_in; i++) {
                       ^
libpd_wrapper/z_libpd.c:154:10: error: 'sys_soundout' undeclared (first use in this function)
   memset(sys_soundout, 0, n_out * sizeof(t_sample));
          ^~~~~~~~~~~~
libpd_wrapper/z_libpd.c:156:24: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   for (p = sys_soundout, i = 0; i < n_out; i++) {
                        ^
libpd_wrapper/z_libpd.c: In function 'libpd_process_short':
libpd_wrapper/z_libpd.c:170:22: error: 'sys_soundin' undeclared (first use in this function)
     for (j = 0, p0 = sys_soundin; j < DEFDACBLKSIZE; j++, p0++) { \
                      ^
libpd_wrapper/z_libpd.c:186:3: note: in expansion of macro 'PROCESS'
   PROCESS(* short_to_sample, * sample_to_short)
   ^~~~~~~
libpd_wrapper/z_libpd.c:171:32: error: 'sys_inchannels' undeclared (first use in this function)
       for (k = 0, p1 = p0; k < sys_inchannels; k++, p1 += DEFDACBLKSIZE) { \
                                ^
libpd_wrapper/z_libpd.c:186:3: note: in expansion of macro 'PROCESS'
   PROCESS(* short_to_sample, * sample_to_short)
   ^~~~~~~
libpd_wrapper/z_libpd.c:175:12: error: 'sys_soundout' undeclared (first use in this function)
     memset(sys_soundout, 0, sys_outchannels*DEFDACBLKSIZE*sizeof(t_sample)); \
            ^
libpd_wrapper/z_libpd.c:186:3: note: in expansion of macro 'PROCESS'
   PROCESS(* short_to_sample, * sample_to_short)
   ^~~~~~~
libpd_wrapper/z_libpd.c:175:29: error: 'sys_outchannels' undeclared (first use in this function)
     memset(sys_soundout, 0, sys_outchannels*DEFDACBLKSIZE*sizeof(t_sample)); \
                             ^
libpd_wrapper/z_libpd.c:186:3: note: in expansion of macro 'PROCESS'
   PROCESS(* short_to_sample, * sample_to_short)
   ^~~~~~~
libpd_wrapper/z_libpd.c: In function 'libpd_process_float':
libpd_wrapper/z_libpd.c:170:22: error: 'sys_soundin' undeclared (first use in this function)
     for (j = 0, p0 = sys_soundin; j < DEFDACBLKSIZE; j++, p0++) { \
                      ^
libpd_wrapper/z_libpd.c:190:3: note: in expansion of macro 'PROCESS'
   PROCESS(,)
   ^~~~~~~
libpd_wrapper/z_libpd.c:171:32: error: 'sys_inchannels' undeclared (first use in this function)
       for (k = 0, p1 = p0; k < sys_inchannels; k++, p1 += DEFDACBLKSIZE) { \
                                ^
libpd_wrapper/z_libpd.c:190:3: note: in expansion of macro 'PROCESS'
   PROCESS(,)
   ^~~~~~~
libpd_wrapper/z_libpd.c:175:12: error: 'sys_soundout' undeclared (first use in this function)
     memset(sys_soundout, 0, sys_outchannels*DEFDACBLKSIZE*sizeof(t_sample)); \
            ^
libpd_wrapper/z_libpd.c:190:3: note: in expansion of macro 'PROCESS'
   PROCESS(,)
   ^~~~~~~
libpd_wrapper/z_libpd.c:175:29: error: 'sys_outchannels' undeclared (first use in this function)
     memset(sys_soundout, 0, sys_outchannels*DEFDACBLKSIZE*sizeof(t_sample)); \
                             ^
libpd_wrapper/z_libpd.c:190:3: note: in expansion of macro 'PROCESS'
   PROCESS(,)
   ^~~~~~~
libpd_wrapper/z_libpd.c: In function 'libpd_process_double':
libpd_wrapper/z_libpd.c:170:22: error: 'sys_soundin' undeclared (first use in this function)
     for (j = 0, p0 = sys_soundin; j < DEFDACBLKSIZE; j++, p0++) { \
                      ^
libpd_wrapper/z_libpd.c:194:3: note: in expansion of macro 'PROCESS'
   PROCESS(,)
   ^~~~~~~
libpd_wrapper/z_libpd.c:171:32: error: 'sys_inchannels' undeclared (first use in this function)
       for (k = 0, p1 = p0; k < sys_inchannels; k++, p1 += DEFDACBLKSIZE) { \
                                ^
libpd_wrapper/z_libpd.c:194:3: note: in expansion of macro 'PROCESS'
   PROCESS(,)
   ^~~~~~~
libpd_wrapper/z_libpd.c:175:12: error: 'sys_soundout' undeclared (first use in this function)
     memset(sys_soundout, 0, sys_outchannels*DEFDACBLKSIZE*sizeof(t_sample)); \
            ^
libpd_wrapper/z_libpd.c:194:3: note: in expansion of macro 'PROCESS'
   PROCESS(,)
   ^~~~~~~
libpd_wrapper/z_libpd.c:175:29: error: 'sys_outchannels' undeclared (first use in this function)
     memset(sys_soundout, 0, sys_outchannels*DEFDACBLKSIZE*sizeof(t_sample)); \
                             ^
libpd_wrapper/z_libpd.c:194:3: note: in expansion of macro 'PROCESS'
   PROCESS(,)
   ^~~~~~~
make: *** [<builtin>: libpd_wrapper/z_libpd.o] Error 1
`

**Comment by danomatika** [comment]
When you changed tag, did you also update the submodules? Those errors look like trying to build with newer Pd sources with an older libpd.

**Comment by residuum** [comment]
The header for all `sys_*` functions are missing.

**Comment by turowskipaul** [comment]
Doh! That was it.  I updated the submodules and built and it's working in Unity now (despite the presence of the input signal an echo—but that's another issue).  Thanks very much guys!!

**Comment by residuum** [comment]
Never use master, when tags are available, master usually is WIP.

**Comment by danomatika** [comment]
Ok, so the issue really came down to building using master and not the stable release. Makes sense as we've not tested any of the new work from the Pd core yet. I will open an issue about this for the next release.

Note, the readme says :

>The master branch contains the latest libpd development and can be considered generally stable. However, we make no guarantees. :)

**Comment by turowskipaul** [comment]
Lesson learned!  You guys rock  :D

**Comment by ATHellboy** [comment]
Hey guys,

I've tried to compile 64 bit version of libpd for Unity3D multiple times but Unity3D crashes after playing scene. I attached 

> out.text

 file for executing 

> mingw64_build_csharp.bat

 file.
[out.txt](https://github.com/libpd/libpd/files/1448715/out.txt)

Also I don't know how to include external libraries of PD in libpd. There is instruction for C language but I didn't find for C#. First, I want to use 

> freeverb~ in my patches.
